### PR TITLE
Test increaseTime behaviour

### PR DESCRIPTION
**Pull request by gpetraglia**
Regarding the discussion on https://github.com/nomiclabs/hardhat/issues/1079:

Make explicit some increaseTime behaviour that were not so clear with
the previous tests.
These behaviour include:
 - If increaseTime is called N times, all of the calls are taken into
account to calculate the new block timestamp
 - If real time elapses, that's also taken into account

The decision to keep this is partially due to the fact that we don't
want to break previously working workflow.

------

There will be another PR that updates the documentation, keeping in mind the needs for documentation PRs

### CODE REVIEW

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts [hunk 1]
```diff
@@ -5,8 +5,10 @@ import VM from "@ethereumjs/vm";
 import { AfterBlockEvent, RunBlockOpts } from "@ethereumjs/vm/dist/runBlock";
 import { assert } from "chai";
 import { Address, BN, bufferToHex } from "ethereumjs-util";
+import { pass } from "fp-ts/lib/Writer";
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:8]: This doesn't seem to be used.

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts [hunk 2]
```diff
@@ -5,8 +5,10 @@ import VM from "@ethereumjs/vm";
...
 import sinon from "sinon";
 
+import { getClosestCallerPackage } from "../../../../internal/util/caller-package";
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:11]: Neither does this.

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts [hunk 3]
```diff
@@ -495,12 +511,29 @@ describe("HardhatNode", () => {
...
+        await assertIncreaseTime(now + delta);
+      });
+
+      it("mining a block having increaseTime called twice counts both calls", async () => {
+        const now = getCurrentTimestamp();
+        const delta = 30;
+        node.increaseTime(new BN(delta));
+        node.increaseTime(new BN(delta));
         await node.mineBlock();
+        await assertIncreaseTime(now + delta * 2);
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:525]: `assertIncreaseTime` calls `mineBlock`, so I guess the previous line should be removed?

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts [hunk 4]
```diff
@@ -495,12 +511,29 @@ describe("HardhatNode", () => {
...
-        const block = await node.getLatestBlock();
-        const blockTimestamp = block.header.timestamp.toNumber();
-        assert.equal(blockTimestamp, now + 30);
+      it("mining a block having called increaseTime takes into account 'real' passing time", async () => {
+        const now = getCurrentTimestamp();
+        const delta = 30;
+        const elapsedTimeInSeconds = 3;
+        node.increaseTime(new BN(delta));
+        clock.tick(elapsedTimeInSeconds * 1_000);
+        await node.mineBlock();
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:534]: Ditto here.

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts (hunk 3 above)
Comment (fvictorio) [reply packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:525]: Actually, I have a slight preference for removing the `mineBlock` call from the `assertIncreaseTime` and having to call it explicitly in the test.

Comment (gonzalpetraglia) [reply packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:525]: Ok, will do that

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts (hunk 1 above)
Comment (gpetraglia) [reply packages/hardhat-core/test/internal/hardhat-network/provider/node.ts]: Vscode might pulled a little sneaky on me. Fixed

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts (hunk 2 above)
Comment (gpetraglia) [reply packages/hardhat-core/test/internal/hardhat-network/provider/node.ts]: Fixed

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts (hunk 3 above)
Comment (gpetraglia) [reply packages/hardhat-core/test/internal/hardhat-network/provider/node.ts:523]: Fixed

File: packages/hardhat-core/test/internal/hardhat-network/provider/node.ts (hunk 4 above)
Comment (gpetraglia) [reply packages/hardhat-core/test/internal/hardhat-network/provider/node.ts]: Fixed

**Review by fvictorio — APPROVED** [review]
Thanks @gonzalpetraglia, I like these tests :rocket:

**Comment by fvictorio** [comment]
Closing this in favor of #1682. I think I can't update this PR because it's created from the master branch of the fork.

**Pull request closed without merging**